porick/views.py

```diff
@@ -189,7 +190,69 @@ def logout():
     flash('Logged out successfully!', 'info')
     return response
 
-
-@app.route('/reset_password')
+@app.route('/reset_password', methods=['GET', 'POST'])
 def reset_password():
-    raise NotImplementedError()
+    if request.method == 'GET':
+        if not request.args.get('key'):
+            return render_template('/pw_reset/request.html')
+
+        token = PasswordReset.query.filter_by(key=request.args['key']).first()
+        if token and token.is_valid:
+            return render_template('/pw_reset/set.html', key=token.key)
+        else:
+            flash('Invalid reset token', 'error')
+            return render_template('/index.html')
+
+
+    elif request.method == 'POST':
+        if not request.args.get('key'):
+        # The user has requested a new token
+            email = request.form['email']
+            user = User.query.filter_by(email=email).first()
+            if not user:
+                flash('Invalid email address provided.', 'error')
+                return render_template('/pw_reset/request.html')
+
+            # If this user already has a valid reset token, don't
+            # let them create a new one until it has expired
+            existing_token = PasswordReset.query.filter_by(user_id=user.id).first()
+            if existing_token:
+                if existing_token.is_valid:
+                    flash('A password reset has already been requested for this user.', 'error')
+                    return render_template('/pw_reset/request.html')
+                else:
+                    db.session.delete(existing_token)
+                    db.session.commit()
+
+            token = PasswordReset(user.id)
+            key = token.key
```

Comment:
> Since it's being dereferenced after the commit

I might be wrong, but I don't believe so. As far as I know, the following is valid:

```
user.name = 'abc'
db.session.commit()
user.name = 'def'
db.session.commit()
```

resulting in `user.name` of `def` being persisted. Can you try it out in the REPL?